miner/proposal_builder.go

```diff
@@ -499,71 +499,181 @@ func (pb *ProposalBuilder) initSharedData(ctx context.Context, current types.Lay
+	if s.prev == 0 {
+		prev, err := ballots.LastInEpoch(pb.db, s.atx, s.epoch)
 		if err != nil && !errors.Is(err, sql.ErrNotFound) {
 			return err
 		}
 		if err == nil {
-			ss.session.prev = prev.Layer
+			s.prev = prev.Layer
 		}
 	}
-	if ss.session.ref == types.EmptyBallotID {
-		ballot, err := ballots.FirstInEpoch(pb.db, ss.session.atx, ss.session.epoch)
+	if s.ref == types.EmptyBallotID {
+		ballot, err := ballots.FirstInEpoch(pb.db, s.atx, s.epoch)
 		if err != nil && !errors.Is(err, sql.ErrNotFound) {
 			return fmt.Errorf("get refballot %w", err)
 		}
 		if errors.Is(err, sql.ErrNotFound) {
-			ss.session.beacon = pb.shared.beacon
-			ss.session.eligibilities.slots = proposals.MustGetNumEligibleSlots(
-				ss.session.atxWeight,
+			s.beacon = pb.shared.beacon
+			s.eligibilities.slots = proposals.MustGetNumEligibleSlots(
+				s.atxWeight,
 				minweight.Select(lid.GetEpoch(), pb.cfg.minActiveSetWeight),
 				pb.shared.active.weight,
 				pb.cfg.layerSize,
 				pb.cfg.layersPerEpoch,
 			)
 		} else {
 			if ballot.EpochData == nil {
-				return fmt.Errorf("atx %d created invalid first ballot", ss.session.atx)
+				return fmt.Errorf("atx %d created invalid first ballot", s.atx)
 			}
-			ss.session.ref = ballot.ID()
-			ss.session.beacon = ballot.EpochData.Beacon
-			ss.session.eligibilities.slots = ballot.EpochData.EligibilityCount
+			s.ref = ballot.ID()
+			s.beacon = ballot.EpochData.Beacon
+			s.eligibilities.slots = ballot.EpochData.EligibilityCount
 		}
 	}
-	if ss.session.eligibilities.proofs == nil {
-		ss.session.eligibilities.proofs = calcEligibilityProofs(
-			ss.signer.VRFSigner(),
-			ss.session.epoch,
-			ss.session.beacon,
-			ss.session.nonce,
-			ss.session.eligibilities.slots,
-			pb.cfg.layersPerEpoch,
-		)
-		ss.log.Info("proposal eligibilities for an epoch", zap.Inline(&ss.session))
-		events.EmitEligibilities(
-			ss.signer.NodeID(),
-			ss.session.epoch,
-			ss.session.beacon,
-			ss.session.atx,
-			uint32(len(pb.shared.active.set)),
-			ss.session.eligibilities.proofs,
+	return nil
+}
+
+func (pb *ProposalBuilder) BuildFor(ctx context.Context,
+	lid types.LayerID,
+	nodeID types.NodeID,
+) (*types.Proposal, types.VRFPostIndex, error) {
+	if err := pb.initSharedData(ctx, lid); err != nil {
+		return nil, 0, err
+	}
+
+	signer := &signerSession{}
+	encodeVotes := func() (*types.Opinion, error) {
+		pb.tortoise.TallyVotes(lid)
+		opinion, err := pb.tortoise.EncodeVotes(ctx, tortoise.EncodeVotesWithCurrent(lid))
+		if err != nil {
+			return nil, fmt.Errorf("encoding votes: %w", err)
+		}
+		return opinion, nil
+	}
+
+	calcMeshHash := func() types.Hash32 {
+		meshHash := pb.decideMeshHash(ctx, lid)
+		return meshHash
+	}
+
+	persistActiveSet := func() error {
+		err := activesets.Add(pb.db, pb.shared.active.id, &types.EpochActiveSet{
+			Epoch: pb.shared.epoch,
+			Set:   pb.shared.active.set,
+		})
+		if err != nil && !errors.Is(err, sql.ErrObjectExists) {
+			return err
+		}
+		return nil
+	}
```

Comment:
Could you inline these? The only reason they were pulled out into the functions was to use the `Once`.